hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TypeUtils.java

```diff
@@ -0,0 +1,328 @@
+            return inlineSchema != null ? new Schema.Parser().parse(inlineSchema) : null;
+        }
+    }
+
+    /**
+     * @param <ID> type of schema identifier
+     * @param <S> type of schema
+     */
+    private abstract static class FieldEnricher<ID, S> {
+        private final int typeKind;
+        private final RelationsStorage relationsStorage;
+
+        FieldEnricher(int typeKind, RelationsStorage relationsStorage) {
+            this.typeKind = typeKind;
+            this.relationsStorage = relationsStorage;
+        }
+
+        void enrich(MappingField field, Map<String, String> mappingOptions, boolean isKey) {
+            String typeName = field.type().getObjectTypeName();
+            field.setType(createFieldType(
+                    field.name().equals(isKey ? QueryPath.KEY : QueryPath.VALUE)
+                            ? () -> getSchemaId(mappingOptions, isKey)
+                            : () -> getFieldSchemaId(getSchema(getSchemaId(mappingOptions, isKey)),
+                                    field.plainExternalName(), typeName),
+                    typeName, new HashMap<>()));
+        }
+
+        protected QueryDataType createFieldType(Supplier<ID> schemaIdSupplier, String typeName,
+                                                Map<String, QueryDataType> seen) {
+            QueryDataType convertedType = seen.get(typeName);
+            if (convertedType != null) {
+                return convertedType;
+            }
+
+            Type type = relationsStorage.getType(typeName);
+            if (type == null) {
+                throw QueryException.error("Encountered type '" + typeName + "', which doesn't exist");
+            }
+
+            ID schemaId = getSchemaId(type.options(), null);
+            if (schemaId == null) {
+                schemaId = schemaIdSupplier.get();
+            }
+
+            S schema = getSchema(schemaId);
+            if (type.getFields().isEmpty()) {
+                type.setFields(resolveFields(schema));
+                relationsStorage.put(typeName, type);
+            }
+
+            convertedType = new QueryDataType(typeName, typeKind);
+            convertedType.setObjectTypeMetadata(getTypeMetadata(schemaId));
+            seen.put(typeName, convertedType);
+
+            for (TypeField field : type.getFields()) {
+                QueryDataType fieldType = field.getQueryDataType();
+                String fieldTypeName = fieldType.getObjectTypeName();
+
+                if (fieldType.isCustomType()) {
+                    fieldType = createFieldType(() -> getFieldSchemaId(schema, field.getName(), fieldTypeName),
+                            fieldTypeName, seen);
+                }
+                convertedType.getObjectFields().add(new QueryDataTypeField(field.getName(), fieldType));
+            }
+
+            return convertedType;
+        }
+
+        protected abstract String getTypeMetadata(ID schemaId);
```

Comment:
could you write some javadoc for these methods, at least what are they supposed to do? Their implementation is sometimes surprising, eg. for avro schema**Id** = full schema.

Reply:
What schema ID and schema mean depend on how we traverse nested schemas for that particular format. For example, for Portable, we traverse `PortableId` → `ClassDefinition` → `FieldDefinition` → `PortableId` ..., whereas for Avro, we traverse `Schema` → `Schema` ... If the traversal cycle consists of a single distinct type, as in Avro, the schema ID and schema have the same type.